User account management for administrators

JoelBts opened the pull request:
**include/check-authorisation.php**
- gave administrators access to new pages user-archive.php, create-user.php, update-user.php
- gave auditors access to user-archive.php

**include/page-defaults.php**
- added function call to update the last active time when a user loads page

**include/utilities.php**
- added a new function updateLastActive which updates a users last active time and location

**pages/create-user.php**
- remade the form for user details

**pages/manage.php**
- shows a list of all users, current and archived
- button to add a new user: redirects to create-user.php
- hovering over a user shows 'edit user?' overlay
- clicking on a user redirects to update-user.php which allows to update that user's details
- archive button: when pressed opens confirmation box to archive user

**pages/update-user.php** (new file)
- allows to modify user details
- retrieves details from the personID from the database and fills in current information
- clicking reset pin button allows the pin to be changed
- auto-generate pin button randomly generates a 4-digit pin and fills it in

**pages/user-archive.php** (new file)
- similar to users.php, but only shows archived users and no button to add new user
- button to return to current users

**pages/users.php**
- displays list of current users and their information
- button to add a new user
- button to view user archive

**scripts/manage-user.js** (new file)
- contains the javascript functions for the user management pages

**scripts/manage.php** (new file)
- contains the php function for adding users to the list on the manage page
- include archived at time if user is archived
- include last active time and location

**scripts/update-sort-order.js** (new file)
- contains the javascript for dynamically updating the users list sort and direction

**scripts/users.php** 
- contains the function for adding users to the users and user-archive pages
- modified function to include last active time and location
- include archived at time if user is archived

**sql/db.sql**
- added columns to the Persons table to keep track of if a user is archived or not and time archived at
- added columns to Persons table to keep track of the time and machine they were last active at

**sql/insert-persons.sql**
- added an archived user
- added lastActiveTime and lastActiveMachineID for existing inserts

**styles/style.css**
- added css for new pages

**system/archive-user.php** (new file)
- contains the php to archive a user
- sets their pin to 'invalid'
- sets isArchived to TRUE and sets archivedAt to the current time

**system/create-user.php**
- changed the redirects after creating a user to go to the users page

**system/login.php**
- modified sql to only get users which are not archived

**system/update-user.php** (new file)
- handles updating the user information to the database

ArranWill commented on scripts/manage-user.js:
You could include a check to see if the randomly generated PIN is already used by someone else in the database.

JoelBts replied:
Yeah, I originally thought this wouldn;t be possible since we're only storing the hashes and can't extract the actual pins to check, but after thinking about it for a while using password_verify the same way as when checking login should work I think. When creating a new user I can also check that the pin isn't already used

ArranWill replied:
Good solution!

ArranWill commented on include/page-defaults.php:
It might be neater to have the conditional statement inside the function. In addition, I'm not sure it's necessary to pass the session and get values as arguments to the function, they should be accessible inside the function.

ArranWill commented on pages/create-user.php:
Missing mysqli_close()?

ArranWill commented on pages/manage.php:
Missing mysqli_free_result()?

ArranWill commented on pages/update-user.php:
Missing mysqli_close()?

ArranWill commented on pages/user-archive.php:
Missing mysqli_free_result()? Could put lines 72-74 in a single block.

ArranWill commented on pages/users.php:
Missing mysqli_free_result().

ArranWill commented on system/archive-user.php:
Missing mysqli_close().

ArranWill commented on system/create-user.php:
Missing mysqli_stmt_close().

ArranWill commented on system/update-user.php:
Missing mysqli_close().

ArranWill reviewed (commented):
Outstanding work overall, you're clearly a very skilled programmer. I've written a few comments, most of which relate to un-freed mysqli resources. In addition, here are some observations from trying out the new pages: 
* There is a possible issue the user may have on the manage page. Typically with a drop-down menu, one clicks the triangle on the right hand side to see the drop-down options, so the user may be surprised when rather than seeing the drop-down options, it reverses the order of the list items. Although the triangle toggle is neat, it might be clearer for the user to have another select drop-down to choose between ascending and descending order. 
* I notice you're passing the machine ID through htmlspecialchars on the user-facing pages, do you think we should implement that for all user-facing pages?
* You could introduce the "active" query string parameter to warn the user when they try to navigate to another tab from the "add user" and "update user pages" (function to make it work may be in another branch)
* It's possible to enter a phone number with too few digits to be valid
* After creating a user there's no feedback from the page that it was successful
* The add user button is present on the users page for auditors
* It is possible to give two users the same PIN, in which case the user with the alphabetically preceding name is logged in
* Upon pressing the reset PIN button on the update user page, it is possible to exit the page without providing a new pin. In this case the PIN is not updated. Maybe the pin input should have the required tag
* The header buttons on the users and manage page are inaccessible on a narrow screen, and the "create user" and "update user" are totally unusable on a narrow screen. This may be a moot point given we haven't been asked to make the system for mobile devices, but I thought I'd mention it. 
* We need to talk as a group about the styling and move towards a consistent approach. One thing I'd like is to have the same margins left and right for the main page elements like the header buttons and the list of users.

JoelBts commented:
Main changes
- Implemented some of Arran's suggestions and fixed the issues he noted
- Removed users page from administrators (set home as manage.php)
- Merged users and manage page for administrators
- users page is only for auditors
- Deleted page user-archive.php as it is now redundant
- Added "show" option on manage and users to show current users, archived users, or all users
- Modified sort-by options to change depending on "show" (i.e. if archived users is selected, archived at is an option)
- Replaced triangle arrow for direction toggle with 'Ascending'/'Descending'
- Restructured users table to show last active time and machine across 2 columns
- Added button to generate a new pin when creating or updating a user
- Checks that a pin doesn't already exist in the database
- Checks that phone number is valid number of digits and starts with 04 or +61
- Archive user button now only shows with edit overlay when user is hovered over
- Added archive user button to update-user page
- Added missing mysqli_free_result and mysqli_stmt_close()
- Added "active=1" to update-user and create-user pages
- Combined all sql insert statements to a single file

Have a look through and see what you think. I'm not too sure about the some of the styling

ArranWill reviewed (approved):
Excellent work implementing all those changes. Some cool stuff going on with the unique PIN validation, well done!
